Add an option to INTO OUTFILE to also print details to standard output.

SmitaRKulkarni opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
resolves https://github.com/ClickHouse/ClickHouse/issues/37490 

Implementation:
- Added a bool to ASTQueryWithOutput & patched the usage in ClientBase.
- Added a new buffer ForkWriteBuffer takes a vector of WriteBuffer and writes data to all of them. It uses the buffer of the first element as its buffer and copies data from first buffer to all the other buffers.
- Added a new bool select_into_outfile_and_stdout in ClientBase to enable/disable progress rendering.

Testing:
 - Added a test tests/queries/0_stateless/02346_into_outfile_and_stdout.sh

Documentation:
- Updated the english documentation for the new option in SELECT.


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

SmitaRKulkarni committed:
Add an option to INTO OUTFILE to also print details to standard output.

Implementation:
- Added a bool to ASTQueryWithOutput & patched the usage in ClientBase.
- Added a new buffer TeeWriteBuffer which extends from WriteBufferFromFile (used to write data to the file) and has WriteBufferFromFileDescriptor (used to write data to stdout). The WriteBufferFromFileDescriptor uses the same buffer as TeeWriteBuffer.
- Added a new bool select_into_outfile_and_stdout in ClientBase to enable/disable progress rendering.

Testing:
 - Added a test tests/queries/0_stateless/02346_into_outfile_and_stdout.sh

Documentation:
- Updated the english documentation for the new option in SELECT.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=39054) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=39054) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=39054) it.</sub>

SmitaRKulkarni committed:
added execute permission on the 02346_into_outfile_and_stdout.sh test file

kssenii reviewed (commented):
LGTM, some small comments

alexey-milovidov commented:
1. Looks like it will write compressed data into stdout if we will ask `INTO OUTFILE 'test.csv.gz' AND STDOUT` - it should not work this way. The tests with compressed files should be added.
2. It is unreasonable to only support stdout and only support WriteBufferFromFile inside TeeWriteBuffer. The name "TeeWriteBuffer" does not imply that it's so specialized and limited.
3. Let's ditch TreeWriteBuffer and implement ForkWriteBuffer which will take an arbitrary number of other WriteBuffers and forward the data to all of them. It can even be used with a single argument (to do nothing useful but it can make the code more generic). The implementation should be generic: it should accept abstract WriteBuffers, not specific ones like WriteBufferFromFile. It should forward operations like `flush` as well.

alexey-milovidov commented on tests/queries/0_stateless/02346_into_outfile_and_stdout.sh:
```diff
@@ -0,0 +1,45 @@
+    local query=$2
+
+    echo "performing test: $test_id"
+    ${CLICKHOUSE_CLIENT} --query "$query" 2>/dev/null
```
Tests should not swallow errors unless it's necessary.

SmitaRKulkarni replied:
Thanks. this has been fixed

alexey-milovidov commented on tests/queries/0_stateless/02346_into_outfile_and_stdout.sh:
```diff
@@ -0,0 +1,45 @@
+    else
+        echo "query failed"
+    fi
+    rm -f "${CLICKHOUSE_TMP}/test_into_outfile_and_stdout_$test_id.out"
```
`${test_id}`
and quotes
for safety.

SmitaRKulkarni replied:
this too has been fixed

SmitaRKulkarni commented:
@alexey-milovidov regarding your comment 
> Looks like it will write compressed data into stdout if we will ask `INTO OUTFILE 'test.csv.gz' AND STDOUT` - it should not work this way. The tests with compressed files should be added.

When compression is used, do we want to store compressed data to file & display plaintext (uncompressed) in STDOUT ? If not, could you please elaborate ?

alexey-milovidov commented:
Yes, exactly this. Uncompressed to stdout, compressed to file.

SmitaRKulkarni committed:
Addressed review comments

Implementation:
- Added a new buffer ForkWriteBuffer takes a vector of WriteBuffer and writes data to all of them. It uses the buffer of the first element as its buffer and copies data from first buffer to all the other buffers
Testing:
- Updated tests/queries/0_stateless/02346_into_outfile_and_stdout.sh
Documentation:
- Updated the english documentation for SELECT.. INTO OUTFILE with AND STDOUT.

SmitaRKulkarni commented on src/IO/TeeWriteBuffer.h:
```diff
@@ -0,0 +1,34 @@
+
+protected:
+    void nextImpl() override;
+    void finalizeImpl() override;
```
TeeWriteBuffer has been removed based on Alexey's comments so this is not needed anymore

SmitaRKulkarni commented on src/IO/TeeWriteBuffer.cpp:
```diff
@@ -0,0 +1,46 @@
+    }
+    catch (Exception &exception)
+    {
+        exception.addMessage("While writing to TeeWriteBuffer ");
```
I have taken this suggestion into account for the new class. thanks

SmitaRKulkarni commented on src/IO/TeeWriteBuffer.cpp:
```diff
@@ -0,0 +1,46 @@
+        stdout_buffer.next();
+        WriteBufferFromFile::nextImpl();
+    }
+    catch (Exception &exception)
```
I have taken this suggestion into account for the new class. thanks

SmitaRKulkarni commented on src/Parsers/ParserQueryWithOutput.cpp:
```diff
@@ -107,6 +107,13 @@ bool ParserQueryWithOutput::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
+        ParserKeyword s_stdout("AND STDOUT");
+        if (s_stdout.ignore(pos, expected))
+        {
+            query_with_output.is_stdout_enabled = true;
```
this has been addressed 👍

SmitaRKulkarni commented on tests/queries/0_stateless/02346_into_outfile_and_stdout.sh:
```diff
@@ -0,0 +1,45 @@
+    if [ "$?" -eq 0 ]; then
+        cat "${CLICKHOUSE_TMP}/test_into_outfile_and_stdout_$test_id.out"
+    else
+        echo "query failed"
```
i have addressed this. please let me know if anything further needs to be addressed

SmitaRKulkarni commented:
@alexey-milovidov thanks for the review,  I have addressed your comments. please have a look.

SmitaRKulkarni committed:
clang tidy feedback addressed: updated src/IO/ForkWriteBuffer.cpp to use const reference

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+{
+    if (sources.empty())
+    {
+        first_buffer = nullptr;
```
this assignment is redundant.

SmitaRKulkarni replied:
addressed through the other comment by removing `first_buffer`

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+    }
+    else
+    {
+        first_buffer = sources.begin()->get();
```
```suggestion
        first_buffer = sources.front().get();
```

SmitaRKulkarni replied:
same here. addressed

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+
+    try
+    {
+        for (const WriteBufferPtr & write_buffer : sources | std::views::reverse)
```
why need `std::views::reverse` here?

kssenii replied:
Ok 👍🏻.
Just my opinion, but probably it could be better to make the this code like this:
```
        auto & source_buffer = sources.front();
        for (auto it = sources.begin() + 1; it != sources.end; ++it)
        {
            auto & buffer = *it;
            buffer->write(source_buffer->buffer().begin(), source_buffer->offset());
            buffer->next();
        }
        source_buffer.next();
```

SmitaRKulkarni replied:
thanks for the suggestion. i have added the same

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+                    while (to_be_written != 0)
+                    {
+                        int bytes_to_copy = std::min(to_be_written, write_buffer->available());
+                        write_buffer->write(first_buffer->buffer().begin()+bytes_written, bytes_to_copy);
```
```suggestion
                        write_buffer->write(first_buffer->buffer().begin() + bytes_written, bytes_to_copy);
```

SmitaRKulkarni replied:
addressed based on the below comment

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+        {
+            if (write_buffer.get() != first_buffer)
+            {
+                //if buffer size if not enough to write, then split the message with buffer length
```
```suggestion
                // If buffer size is not enough for write, then split the message with buffer length.
```

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+            if (write_buffer.get() != first_buffer)
+            {
+                //if buffer size if not enough to write, then split the message with buffer length
+                if (write_buffer->available() < first_buffer->offset())
+                {
+                    size_t bytes_written = 0;
+                    auto to_be_written = first_buffer->offset();
+                    while (to_be_written != 0)
+                    {
+                        int bytes_to_copy = std::min(to_be_written, write_buffer->available());
+                        write_buffer->write(first_buffer->buffer().begin()+bytes_written, bytes_to_copy);
+                        write_buffer->next();
+                        bytes_written += bytes_to_copy;
+                        to_be_written -= bytes_to_copy;
+                    }
+                }
```
You do not really need this code, simply `write_buffer->write(first_buffer->buffer().begin(), first_buffer->offset());` should also work ok. If the buffer length is smaller, then it is ok, because there is `nextIfAtEnd()` in `write()` method.

SmitaRKulkarni replied:
yes good point, this is not needed. I have addressed it

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+
+void ForkWriteBuffer::nextImpl()
+{
+    if (!first_buffer)
+        return;
```
If we make sure that this buffer cannot be created with an empty set of buffers (there is exception in constructor now), then this check is not needed.

kssenii replied:
Also probably there is no point to have a separate variable `first_buffer` as it always equals `sources.front()`. In my opinion it adds redundant complexity with worrying that `first_buffer`, which is a raw pointer, is valid.

SmitaRKulkarni replied:
good catch! i have removed `first_buffer`

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,83 @@
+
+}
+
+void ForkWriteBuffer::finalizeImpl()
+{
+    next();
+}
```
Probably no point to override this method as its default implementation in WriteBuffer is exactly the same as written here.

SmitaRKulkarni committed:
Addressed review comments
- Removed first_buffer, finalizeImpl and logic to different buffer size from ForkWriteBuffer

SmitaRKulkarni committed:
addressed review comment: ForkWriteBuffer for loop

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,52 @@
+{
+    if (sources.empty())
+    {
+        throw Exception("ForkWriteBuffer required WriteBuffer is not provided", ErrorCodes::CANNOT_CREATE_IO_BUFFER);
```
```suggestion
        throw Exception("Expected non-zero number of buffers for `ForkWriteBuffer`", ErrorCodes::CANNOT_CREATE_IO_BUFFER);
```

SmitaRKulkarni replied:
i have changed this as per your suggestion

kssenii commented on tests/queries/0_stateless/02346_into_outfile_and_stdout.sh:
```diff
@@ -0,0 +1,75 @@
+
+
+
+perform "select" "SELECT 1, 2, 3 INTO OUTFILE '${CLICKHOUSE_TMP}/test_into_outfile_and_stdout_select.out' AND STDOUT"
```
Let's also write a test for the case when we insert more than one buffer size, e.g. `nextImpl()` should be called more than once.

SmitaRKulkarni replied:
Based on my understanding of your comment I have added union test case which calls nextImpl() twice. Hope this addresses it.

SmitaRKulkarni committed:
Addressed review comments
- Updated exception message in ForkWriteBuffer
- Added test case to tests/queries/0_stateless/02346_into_outfile_and_stdout.sh for calling nextImpl more than once

alexey-milovidov commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,52 @@
+
+ForkWriteBuffer::~ForkWriteBuffer()
+{
+    finalize();
```
I see that there is no overloaded finalize method in ForkWriteBuffer, does it make sense to call finalize here?

SmitaRKulkarni replied:
I have called finalize() here so that we call the nextImpl() of ForkWriteBuffer and flush data to the buffers.

kssenii replied:
Probably makes sense to avoid calling `write` for each buffer, but only call `next` (now as default implementation of `finalize` is `finalize() -> finalizeImpl() -> next()`, so `nextImpl` of `ForkWriteBuffer` is called (as Smita mentioned) and we have `write` method also called before `next` for each of the buffers.

SmitaRKulkarni replied:
If we just want to call next method (without write), then I will remove the `finalize()` call from here. The destructor of each WriteBuffer will call the `finalize()->finalizeImpl()->next()`. Please let me know your thoughts.

kssenii replied:
But in general we do not know that all write buffers call `finalize` in destructor (even though it we try to do it everywhere). So if we remove `finalize()` call from here, then for example what if someone passes `ForkWriteBuffer` in buffers list to another `ForkWriteBuffer`, then `next()` will not be called. E.g. we cannot rely that all write buffers indeed call `finalize()->finalizeImpl()->next()` in destructor.

alexey-milovidov commented on src/Client/ClientBase.cpp:
```diff
@@ -549,11 +549,26 @@ try
-                    compression_method,
-                    compression_level
-                );
+                if (query_with_output->is_into_outfile_with_stdout)
```
Ok, but maybe we can avoid duplication?
Create the main write buffer, and if needed - move it to the ForkWriteBuffer?

SmitaRKulkarni replied:
i have addressed this. thanks

alexey-milovidov commented:
Everything else LGTM.

SmitaRKulkarni committed:
Addressed review comment
- updated creating ForkWriteBuffer in ClientBase to avoid duplication

SmitaRKulkarni committed:
Addressed review comments
- Updated destructor of ForkWriteBuffer to clear buffer

kssenii commented on src/IO/ForkWriteBuffer.cpp:
```diff
@@ -0,0 +1,53 @@
+ForkWriteBuffer::~ForkWriteBuffer()
+{
+    finalize();
+    set(nullptr, 0);
```
I think this `set` is not needed

SmitaRKulkarni replied:
From my understanding, after `~ForkWriteBuffer()`, the `~()WriteBuffer()-> finalize()->next()` will be called. In which `offset() `is checked and if it's not zero then `nextImp()` of WriteBuffer is called. So I added this to clear the Buffer so that we dont call nextImpl(). If we were to remove the finalize() call from above comment, then we would need to clear buffer / set `pos = working_buffer.begin()`. (currently `next() ` call always clears it)
All of this is assuming we are in ForkWriteBuffer destructor with some data in buffer yet to be written. Please let me know your thoughts

kssenii replied:
> From my understanding, after ~ForkWriteBuffer(), the ~()WriteBuffer()-> finalize()->next() will be called.

No, because the destructor of `WriteBuffer` is defined as `virtual ~WriteBuffer() = default;`.

SmitaRKulkarni replied:
Oh thats a good point. Silly of me to have missed it. 
I have addressed it.

SmitaRKulkarni committed:
Addressed review comments
- Added finalizeImpl() override in ForkWriteBuffer to call finalize() of all the buffers.
- Removed clearing buffer in ForkWriteBuffer destructor.

kssenii commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into add-option-for-outfile-to-print-stdout

alesapin commented:
Failures are unrelated.
